From 183b2099d7f7eb193e71991ae9a9f48a1411cbd4 Mon Sep 17 00:00:00 2001
From: jaimergp <jaimergp@users.noreply.github.com>
Date: Mon, 14 Aug 2023 15:40:57 +0200
Subject: [PATCH] Install napari from repository in docker image (#6097)
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Closes https://github.com/napari/napari/issues/6096

We have a workflow that runs on each push and publishes a docker image
to ghcr.io. This includes tagged releases.

However, the Dockerfile installs napari from PyPI, which will always be
lagging. Hence, the docker container for 0.4.18 ships 0.4.17... and so
do all the previous commits since 0.4.17 excluded ðŸ˜‚ We have been
publishing the same Docker image for each push for a while and maaaybe
we should delete a few (although layers are cached, so not so bad).

- `build-args`:
https://github.com/docker/build-push-action/tree/v4/#inputs
- ARG vs ENV:
https://vsupalov.com/docker-arg-env-variable-guide/#arg-and-env-availability

- [x] Bug-fix (non-breaking change which fixes an issue)

- [x] CI passes and reports correct version

- [X] My PR is the minimum possible work for the desired functionality
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added tests that prove my fix is effective or that my
feature works
---
 .github/workflows/docker-singularity-publish.yml | 12 ++++++++++--
 dockerfile                                       |  6 ++++--
 2 files changed, 14 insertions(+), 4 deletions(-)

diff --git a/.github/workflows/docker-singularity-publish.yml b/.github/workflows/docker-singularity-publish.yml
index 1ea9c815..bf3838c4 100644
--- a/.github/workflows/docker-singularity-publish.yml
+++ b/.github/workflows/docker-singularity-publish.yml
@@ -77,7 +77,8 @@ jobs:
       # Build and push Docker image with Buildx (don't push on PR)
       # https://github.com/docker/build-push-action
       - name: Build and push Docker image
-        uses: docker/build-push-action@v3
+        uses: docker/build-push-action@v4
+        id: docker_build
         with:
           context: .
           push: ${{ github.event_name != 'pull_request' }}
@@ -85,7 +86,14 @@ jobs:
           labels: ${{ steps.meta.outputs.labels }}
           file: "dockerfile"
           target: ${{ matrix.target }}
-
+          # We build from the the tag name if triggered by tag, otherwise from the commit hash
+          build-args: |
+            NAPARI_COMMIT=${{ github.ref_type == 'tag' && github.ref_name || github.sha }}
+
+      - name: Test Docker image
+        run: |
+          docker run --rm --entrypoint=/bin/bash ${{ steps.docker_build.outputs.imageid }} -ec "python3 -m napari --version"
+
 # ----

   build2:
diff --git a/dockerfile b/dockerfile
index 006eb759..07873c9d 100644
--- a/dockerfile
+++ b/dockerfile
@@ -6,6 +6,7 @@ FROM --platform=linux/amd64 ubuntu:22.04 AS napari
 # below env var required to install libglib2.0-0 non-interactively
 ENV TZ=America/Los_Angeles
 ARG DEBIAN_FRONTEND=noninteractive
+ARG NAPARI_COMMIT=main

 # install python resources + graphical libraries used by qt and vispy
 RUN apt-get update && \
@@ -33,8 +34,9 @@ RUN apt-get update && \
         libxcb-shape0 \
         && apt-get clean

-# install napari release version
-RUN pip3 install napari[all]
+# install napari from repo
+# see https://github.com/pypa/pip/issues/6548#issuecomment-498615461 for syntax
+RUN pip3 install "napari[all] @ git+https://github.com/napari/napari.git@${NAPARI_COMMIT}"

 # copy examples
 COPY examples /tmp/examples
--
2.34.1
